silence/silence.go

```diff
@@ -0,0 +1,792 @@
+	g.mtx.Lock()
+	defer g.mtx.Unlock()
+
+	if delta := g.st.mergeDelta(gd); len(delta) > 0 {
+		return delta, nil
+	}
+	return nil, nil
+}
+
+// OnGossipBroadcast implements the mesh.Gossiper interface.
+func (g gossiper) OnGossipBroadcast(src mesh.PeerName, msg []byte) (mesh.GossipData, error) {
+	gd, err := decodeGossipData(msg)
+	if err != nil {
+		return nil, err
+	}
+	g.mtx.Lock()
+	defer g.mtx.Unlock()
+
+	return g.st.mergeDelta(gd), nil
+}
+
+// OnGossipUnicast implements the mesh.Gossiper interface.
+// It always panics.
+func (g gossiper) OnGossipUnicast(src mesh.PeerName, msg []byte) error {
+	panic("not implemented")
+}
+
+type gossipData map[string]*pb.MeshSilence
+
+func decodeGossipData(msg []byte) (gossipData, error) {
+	gd := gossipData{}
+	rd := bytes.NewReader(msg)
+
+	for {
+		var s pb.MeshSilence
+		if _, err := pbutil.ReadDelimited(rd, &s); err != nil {
+			if err == io.EOF {
+				break
+			}
+			return gd, err
+		}
+		gd[s.Silence.Id] = &s
+	}
+	return gd, nil
+}
+
+// Encode implements the mesh.GossipData interface.
+func (gd gossipData) Encode() [][]byte {
+	// Split into sub-messages of ~1MB.
+	const maxSize = 1024 * 1024
+
+	var (
+		buf bytes.Buffer
+		res [][]byte
+		n   int
+	)
+	for _, s := range gd {
+		m, err := pbutil.WriteDelimited(&buf, s)
+		n += m
+		if err != nil {
+			// TODO(fabxc): log error and skip entry. Or can this really not happen with a bytes.Buffer?
+			panic(err)
+		}
+		if n > maxSize {
+			res = append(res, buf.Bytes())
+			buf = bytes.Buffer{}
+		}
+	}
+	if buf.Len() > 0 {
+		res = append(res, buf.Bytes())
+	}
+	return res
+}
+
+func (gd gossipData) clone() gossipData {
+	res := make(gossipData, len(gd))
+	for id, s := range gd {
+		res[id] = s
+	}
+	return res
+}
+
+// Merge the silence set with gosip data and reutrn a new silence state.
+func (gd gossipData) Merge(other mesh.GossipData) mesh.GossipData {
+	for id, s := range other.(gossipData) {
+		prev, ok := gd[id]
+		if !ok {
+			gd[id] = s
+			continue
+		}
+		pts, err := ptypes.Timestamp(prev.Silence.UpdatedAt)
+		if err != nil {
+			panic(err)
+		}
+		sts, err := ptypes.Timestamp(s.Silence.UpdatedAt)
+		if err != nil {
+			panic(err)
+		}
+		if pts.Before(sts) {
+			gd[id] = s
+		}
+	}
+	return gd
+}
+
+// mergeDelta behaves like Merge but returns a gossipData only
+// containing things that have changed.
+func (gd gossipData) mergeDelta(od gossipData) gossipData {
+	delta := gossipData{}
+	for id, s := range od {
+		prev, ok := gd[id]
+		if !ok {
+			gd[id] = s
+			delta[id] = s
+			continue
+		}
+		pts, err := ptypes.Timestamp(prev.Silence.UpdatedAt)
+		if err != nil {
+			panic(err)
+		}
+		sts, err := ptypes.Timestamp(s.Silence.UpdatedAt)
+		if err != nil {
+			panic(err)
+		}
+		if pts.Before(sts) {
+			gd[id] = s
+			delta[id] = s
+		}
+	}
+	return delta
+}
+
+// replaceFile wraps a file that is moved to another filename on closing.
+type replaceFile struct {
+	*os.File
+	filename string
+}
+
+func (f *replaceFile) Close() error {
```

Comment:
This is copied code that's tested in `nflog` and will be factored out into a util package. Will add docs there.